### PR TITLE
nrfx_pwm: Make possible to skip GPIO configuration in nrfx_pwm_init()

### PR DESCRIPTION
**Pull request by zyan-wu**
We need to configure gpio pins differently from its default values while initiate pwm and pwm doesn't allow us. The new macro is added so that a user can configure the pins before invoking nrfx_init_pwm if NRFX_PWM_USER_CONFIG_GPIO is set to non-zero.

### CODE REVIEW

**Comment by zyan-wu** [comment]
Hey @kl-cruz ,

I assume you are the code owner of this project. Do you have some time to look at my PR?

**Comment by Xenoamor** [comment]
Can't you just init the PWM and then change the pin configuration before enabling it?

**Comment by zyan-wu** [comment]
> 
> 
> Can't you just init the PWM and then change the pin configuration before enabling it?

@Xenoamor , it could be extremely hard for us to do so. Here are the reasons.
- In this case, we need to drive the output as H0H1 to control our ADC module. H0H1 is what our HW engineer requires and the default value is S0S1. So I probably have to ask him if a short period of S0S1 at the startup cause any trouble. Meanwhile we also set the line high to init our ADC module before we init pwm. H1 -> S1 -> H1 might cause trouble. So we have to discuss in great details with HW engineer. I would rather not follow what my HW engineer asks us to do.
- Meanwhile, we need to do the similar thing to ppi, spi, twi etc. Discussing with HW engineer if we can tolerate a short period of mis-configuration would be hard and time-consuming. So this PR is used to discuss with you a workable solution for us. Re-configuration after init is not feasible.
- And also, even if we figured out that your init configuration is the same or won't cause problems, how could we know your future changes won't cause problems to us. It is also impossible for us to go through the process every time we want to upgrade nrfx. We'd rather configure those pins ourselves at init().

**Comment by zyan-wu** [comment]
@kl-cruz @Xenoamor , any idea?

**Comment by Xenoamor** [comment]
Perhaps `nrfx_pwm_config_t` should be extended to include the full range of pin options?

**Comment by zyan-wu** [comment]
It is a possible solution. How do you like it to be implemented?

I can add a few fields to nrfx_pwm_config_t like the following. However, how can it be compatible with the old code? Especially we cannot assume where users initiate nrfx_pwx_config_t so those variables could be any uninitialized values. 
`typedef struct
{
nrf_gpio_pin_input_t inputs[NRF_PWM_CHANNEL_COUNT];
nrf_gpio_pin_pull_t pulls[NRF_PWM_CHANNEL_COUNT];
nrf_gpio_pin_drive_t drives[NRF_PWM_CHANNEL_COUNT];
nrf_gpio_pin_sense_t senses[NRF_PWM_CHANNEL_COUNT];
}

`
Still I can add a macro to protect those fields.  
And to use it, it becomes.
`
nrfx_pwm_config_t c;
c.inputs[0] =  NRF_GPIO_PIN_INPUT_DISCONNECT;
c.pulls[0] = NRF_GPIO_PIN_NOPULL;
c.drives[0] = NRF_GPIO_PIN_H0H1;
c.senses[0] = NRF_GPIO_PIN_NOSENSE;
`
It seems cumbersome compared to how I did my change. Meanwhile I can do the following to all of my pins in one place.

`nrf_gpio_cfg(PIN,              NRF_GPIO_PIN_DIR_OUTPUT, NRF_GPIO_PIN_INPUT_DISCONNECT, NRF_GPIO_PIN_NOPULL,   NRF_GPIO_PIN_H0H1, NRF_GPIO_PIN_NOSENSE);
`

**Comment by zyan-wu** [comment]
@Xenoamor any comments?

**Comment by Xenoamor** [comment]
Sorry I can't comment. I'm not a maintainer just a user
It would likely require an API/SDK major version change as like you say it may not be backwards compatiable

**Comment by anangl** [comment]
@zyan-wu Thanks for your contribution. I agree that it should be possible to prevent the PWM driver from modifying the GPIO configuration of the pins used as output channels. However, I think that instead of adding an nrfx_config macro for this, it would be better to add a new field to the `nrfx_pwm_config_t` structure (`skip_gpio_configuration` would be probably a proper name for it). This way it will be possible to configure this per PWM instance, not globally.
@kl-cruz What you think?

**Comment by kl-cruz** [comment]
@anangl I agree with you. Exapanding current API by skipping GPIO configuration in the ```nrfx_pwm_config_t``` structure seems to be the best approach. Remember to add new field to default config: 
https://github.com/NordicSemiconductor/nrfx/blob/master/drivers/include/nrfx_pwm.h#L117

**Comment by zyan-wu** [comment]
@kl-cruz , sure, I'll try this and update the PR.

**Comment by zyan-wu** [comment]
@kl-cruz @anangl , I've updated the PR. Please let me know if it is what you want.

**Comment by zyan-wu** [comment]
@kl-cruz @anangl , any comments?

File: drivers/include/nrfx_pwm.h [hunk 1]
```diff
@@ -94,12 +94,13 @@ typedef struct
...
-    uint16_t           top_value;    ///< Value up to which the pulse generator counter counts.
-    nrf_pwm_dec_load_t load_mode;    ///< Mode of loading sequence data from RAM.
-    nrf_pwm_dec_step_t step_mode;    ///< Mode of advancing the active sequence.
+    uint8_t            irq_priority;  ///< Interrupt priority.
+    nrf_pwm_clk_t      base_clock;    ///< Base clock frequency.
+    nrf_pwm_mode_t     count_mode;    ///< Operating mode of the pulse generator counter.
+    uint16_t           top_value;     ///< Value up to which the pulse generator counter counts.
+    nrf_pwm_dec_load_t load_mode;     ///< Mode of loading sequence data from RAM.
+    nrf_pwm_dec_step_t step_mode;     ///< Mode of advancing the active sequence.
+    bool               skip_gpio_cfg; ///< Skip configuration of gpio if it is true.
```
Comment (kl-cruz) [inline drivers/include/nrfx_pwm.h]: I'd rather to rephrase it to shorter version:
```Skip the GPIO configuration.```

File: drivers/src/nrfx_pwm.c [hunk 2]
```diff
@@ -92,6 +92,11 @@ static void configure_pins(nrfx_pwm_t const *        p_instance,
             bool inverted = output_pin &  NRFX_PWM_PIN_INVERTED;
             out_pins[i]   = output_pin & ~NRFX_PWM_PIN_INVERTED;
 
+            if (p_config->skip_gpio_cfg)
+            {
+                continue;
```
Comment (kl-cruz) [inline drivers/src/nrfx_pwm.c]: I'd prefer to do not use continue as a point of exit from current context. 
```
    if (output_pin != NRFX_PWM_PIN_NOT_USED)
    {
        out_pins[i] = output_pin & ~NRFX_PWM_PIN_INVERTED;

        if (!p_config->skip_gpio_cfg)
        {
            nrf_gpio_pin_write(out_pins[i], output_pin & NRFX_PWM_PIN_INVERTED);
            nrf_gpio_cfg_output(out_pins[i]);
        }
    }
    else
    {
    ...
```

Comment (anangl) [reply drivers/src/nrfx_pwm.c]: I second that. Use of `continue` obscures the control-flow here.

File: drivers/include/nrfx_pwm.h (hunk 1 above)
Comment (anangl) [reply drivers/include/nrfx_pwm.h]: Well, I think the description should be actually more detailed. Something like:
```
    bool               skip_gpio_cfg; ///< Skip the GPIO configuration.
                                      /**< When this flag is set, the user is responsible for
                                       *   providing the proper configuration of the output pins,
                                       *   as the driver does not touch it at all. */
```

Comment (zyan-wu) [reply drivers/include/nrfx_pwm.h]: I've updated the comments according to your description.

File: drivers/src/nrfx_pwm.c (hunk 2 above)
Comment (zyan-wu) [reply drivers/src/nrfx_pwm.c]: I use the suggested logic.

File: drivers/include/nrfx_pwm.h (hunk 1 above)
Comment (zyan-wu) [reply drivers/include/nrfx_pwm.h]: I will squash the commits and I've updated the commit message.

File: drivers/include/nrfx_pwm.h [hunk 3]
```diff
@@ -117,19 +118,20 @@ typedef struct
...
+                       _out_2,                                    \
+                       _out_3                                     \
+                     },                                           \
+    .irq_priority  = NRFX_PWM_DEFAULT_CONFIG_IRQ_PRIORITY,        \
+    .base_clock    = NRF_PWM_CLK_1MHz,                            \
+    .count_mode    = NRF_PWM_MODE_UP,                             \
+    .top_value     = 1000,                                        \
+    .load_mode     = NRF_PWM_LOAD_COMMON,                         \
+    .step_mode     = NRF_PWM_STEP_AUTO,                           \
+    .skip_gpio_cfg = false,                                       \
```
[the remark this answers is not in the document]
Comment (zyan-wu) [reply drivers/include/nrfx_pwm.h]: done.

**Comment by kl-cruz** [comment]
Hi @zyan-wu , code looks good for me now. Thank you!
You need to squash all the commits together and push one commit using force option. Please let me know if you need help. 
Example reference:
http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html

**Comment by zyan-wu** [comment]
> 
> 
> Hi @zyan-wu , code looks good for me now. Thank you!
> You need to squash all the commits together and push one commit using force option. Please let me know if you need help.
> Example reference:
> http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html

@kl-cruz , done, things are squashed into one commit already.
